source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,423 @@
+  }
+
+  // Finalize this registry. No further changes are allowed after this point. This guaranteed that
+  // all map created by the process have the same variable size and custom registrations. This
+  // function could be called multiple times safely and only the first call will have effect.
+  void finalize() { finalized_ = true; }
+
+  template <class T> InlineMapPtr<T> createInlineMap() {
+    ASSERT(finalized_, "Cannot create inline map before finalized()");
+
+    // Call finalize() anyway to make sure that the registry is finalized and no any new inline
+    // keys could be registered.
+    finalize();
+    return InlineMap<T>::create(*this);
+  }
+
+#ifndef NDEBUG
+public:
+  // This is the registry id that the inline map created by. This is used to validate the inline
+  // key validity.
+  uint64_t registryId() const { return registry_id_; }
+
+private:
+  static uint64_t generateRegistryId() {
+    static std::atomic<uint64_t> next_id{0};
+    return next_id++;
+  }
+
+  uint64_t registry_id_{generateRegistryId()};
+#endif
+
+private:
+  bool finalized_{};
+  RegistrationSet registration_set_;
+  RegistrationMap registration_map_;
+};
+
+/**
+ * Helper class to get inline map registry singleton based on the scope. If cross-modules inline map
+ * registry is necessary, this class should be used.
+ */
+class InlineMapRegistryHelper {
+public:
+  using InlineKey = InlineMapRegistry<std::string>::InlineKey;
+  using ManagedRegistries = absl::flat_hash_map<std::string, InlineMapRegistry<std::string>*>;
+
+  /**
+   * Get the inline map registry singleton based on the scope.
+   */
+  template <class Scope>
+  static InlineMapRegistry<std::string>::InlineKey registerInlinKey(absl::string_view key) {
+    return mutableRegistry<Scope>().registerInlineKey(key);
+  }
+
+  /**
+   * Get registered inline key by the key name.
+   */
+  template <class Scope>
+  static absl::optional<InlineMapRegistry<std::string>::InlineKey>
+  getInlineKey(absl::string_view key) {
+    return mutableRegistry<Scope>().getInlineKey(key);
+  }
+
+  /**
+   * Create inline map based on the scope.
+   */
+  template <class Scope, class T>
+  static InlineMapRegistry<std::string>::InlineMapPtr<T> createInlineMap() {
+    return mutableRegistry<Scope>().template createInlineMap<T>();
+  }
+
+  /**
+   * Get all registries managed by this helper. This this helpful to finalize all registries and
+   * print the debug info of inline map registry.
+   */
+  static const ManagedRegistries& registries() { return mutableRegistries(); }
+
+  /**
+   * Finalize all registries managed by this helper. This should be called after all inline keys
+   * are registered.
+   */
+  static void finalize() {
+    for (auto& registry : mutableRegistries()) {
+      registry.second->finalize();
+    }
+  }
+
+private:
+  template <class Scope> static InlineMapRegistry<std::string>& mutableRegistry() {
+    static InlineMapRegistry<std::string>* registry = []() {
```

Comment:
again, let's let the caller make the determination of memory policy; don't do it in the class implementation.

And if you do add an optional construct_on_first_use mechanism just use CONSTRUCT_ON_FIRST_USE or MUTABLE_CONSTRUCT_ON_FIRST_USE.

Reply:
This is not part of inline map library. It's a optional helper about using the inline map library. 
And I think CONSTRUCT_ON_FIRST_USE or MUTABLE_CONSTRUCT_ON_FIRST_USE cannot support to create a object by a lambda?

Reply:
That's probably fine, but can we have it this layer that's got statics in it in a separate file? That will make it easier to review separately from the pure data structure.